bindings/go/blob/compression/blob.go

```diff
@@ -0,0 +1,202 @@
+// The base blob will be compressed using the canonical compression method (GZIP).
+func Compress(b blob.ReadOnlyBlob) *Blob {
+	return &Blob{ReadOnlyBlob: b, CompressionMethod: MethodCanonical}
+}
+
+// Blob represents a compressed blob that wraps a base ReadOnlyBlob.
+// It implements the MediaTypeAware interface and provides compression functionality.
+type Blob struct {
+	blob.ReadOnlyBlob
+	CompressionMethod Method
+}
+
+// MediaType returns the media type of the compressed blob.
+// It appends the appropriate compression suffix to the base blob's media type.
+// Returns the media type and true if the media type is known.
+func (b *Blob) MediaType() (mediaType string, known bool) {
+	return mediaTypeForBlob(b.ReadOnlyBlob, b.CompressionMethod), true
+}
+
+// mediaTypeForBlob determines the media type for a blob based on its compression method.
+// It handles different compression methods and returns the appropriate media type string.
+func mediaTypeForBlob(b blob.ReadOnlyBlob, method Method) string {
+	var mediaType string
+	switch method {
+	case MethodGzip:
+		fallthrough
+	default:
+		mediaType = getMediaType(b, MediaTypeGzipSuffix, MediaTypeGzip)
+	}
+	return mediaType
+}
+
+// getMediaType determines the media type for a blob, considering its compression.
+// If the blob implements MediaTypeAware, it uses that media type and appends the compression suffix.
+// Otherwise, it returns the default media type for the compression method.
+func getMediaType(b blob.ReadOnlyBlob, ext, def string) string {
+	var mediaType string
+	if mediaTypeAware, ok := b.(blob.MediaTypeAware); ok {
+		if mediaType, ok = mediaTypeAware.MediaType(); ok {
+			mediaType += ext
+		}
+	}
+	if mediaType == "" {
+		mediaType = def
+	}
+	return mediaType
+}
+
+// ReadCloser returns an io.ReadCloser that provides access to the compressed blob data.
+// The returned reader will automatically compress the data from the base blob using the specified compression method.
+// The caller is responsible for closing the returned ReadCloser.
+func (b *Blob) ReadCloser() (io.ReadCloser, error) {
+	base, err := b.ReadOnlyBlob.ReadCloser()
+	if err != nil {
+		return nil, err
+	}
+
+	reader, writer := io.Pipe()
+
+	go compress(base, writer, b.CompressionMethod)
+
+	return reader, nil
+}
+
+// compress compresses the data from the reader with the specified compression method and writes it to the writer.
+func compress(reader io.ReadCloser, writer *io.PipeWriter, method Method) {
+	var compressed io.WriteCloser
+	switch method {
+	case MethodGzip:
+		fallthrough
+	default:
+		compressed = gzip.NewWriter(writer)
+	}
+
+	_, err := io.Copy(compressed, reader)
+	writer.CloseWithError(errors.Join(err, compressed.Close(), reader.Close()))
+}
+
+// Decompress creates a decompressed version of the given blob if it is compressed.
+// It detects the compression method based on the blob's media type and returns a new
```

Comment:
Why do we only decompress based on the media type?

We could also use something like [`DetectCompressionFormat`](https://pkg.go.dev/github.com/containers/image/v5/pkg/compression@v5.34.3#DetectCompressionFormat) to get the compression format.

Reply:
we could, but tbh im not sure if we should do that. if the media type is unknown, would you prefer that?

Reply:
I think there are good reasons to rely on the mediaType and enforce that it is set. So, I am fine with the current approach.